doc: getting_started: Add instructions for multi-platform Zephyr SDK

stephanosio opened the pull request:
```
This commit adds the Linux, macOS and Windows instructions for
installing the new multi-platform Zephyr SDK.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>
```

[Zephyr SDK 0.14.0](https://github.com/zephyrproject-rtos/sdk-ng/releases/tag/v0.14.0).

stephanosio committed:
doc: getting_started: Add wget as a common host dependency

This commit adds the instruction for installing the `wget` application
for macOS and Windows hosts, such that it is available on all three
host platforms when the Getting Started Guide is followed.

The rationale behind this is as follows:

  * The Zephyr documentations, including the Getting Started Guide,
    makes extensive use of the wget command.

  * wget is a purpose-made tool for get/download requests and is more
    user-friendly than curl for the purpose of downloading files.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio committed:
doc: getting_started: Add instructions for multi-platform Zephyr SDK

This commit adds the Linux, macOS and Windows instructions for
installing the new multi-platform Zephyr SDK.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

mbolivar-nordic commented on doc/getting_started/index.rst:
```diff
@@ -471,71 +471,142 @@ Install a Toolchain
 A toolchain provides a compiler, assembler, linker, and other programs required
 to build Zephyr applications.
 
+The Zephyr Software Development Kit (SDK) contains toolchains for each of
```
I thought we were going to stop calling this the SDK? Did I miss a memo?

stephanosio replied:
As per the discussion in the TSC, we will stick with the `Zephyr SDK` nomenclature for now.

mbolivar-nordic reviewed (commented):
Looks good overall. I don't understand the naming, though. I thought we discussed at the TSC that this was going to be renamed from "Zephyr SDK", because there was too much potential for confusion in the embedded space with "vendor SDK", which includes all the firmware libraries etc for developing on an SoC. Did I misunderstand something?

stephanosio commented:
Ready for review. DNM until Zephyr SDK 0.14.0 is released.

stephanosio commented:
[Zephyr SDK 0.14.0 has been released](https://github.com/zephyrproject-rtos/sdk-ng/releases/tag/v0.14.0).

Ready for final review and merging.

stephanosio committed:
doc: beyong-GSG: Update instructions for multi-platform Zephyr SDK

Since Zephyr SDK is now supported on all major operating systems, there
is no need to restrict it to "on Linux."

This commit also removes an endorsement of the GNU Arm Embedded
toolchain because there is no need to recommend and/or use it anymore
for the aforementioned reason.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio commented:
@carlescufi @gmarull @mbolivar-nordic Could you review this so that we can merge it and announce the SDK 0.14.0 release?

gmarull commented on doc/getting_started/index.rst:
```diff
@@ -471,71 +471,142 @@ Install a Toolchain
+         .. code-block:: bash
+
+            cd ~
+            wget https://github.com/zephyrproject-rtos/sdk-ng/releases/download/v0.14.0/zephyr-sdk-0.14.0_macos-x86_64.tar.gz
```
maybe mention that ARM macOS users should pick `aarch64`?

carlescufi replied:
Yes please

stephanosio replied:
done

gmarull reviewed (dismissed):
LGTM, one minor comment

carlescufi commented on doc/getting_started/index.rst:
```diff
@@ -471,71 +471,142 @@ Install a Toolchain
 
          .. note::
-            It is recommended to install the Zephyr SDK at one of the following locations:
+            It is recommended to extract the Zephyr SDK bundle at one of the following locations:
```
why did you reduce the list from the original one? eg. `~/.local/` is not there any longer?

carlescufi replied:
Applies to macOS too

mbolivar-nordic replied:
It is a standard practice: https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html

stephanosio replied:
It is questionable whether that loosely defined one page doc from the XDG qualifies as a "standard practice" and not just "XDG practice," but I will add these back since we would not want to delay getting this PR merged any longer.

stephanosio replied:
As for `/usr`, I will remove it since it is outright wrong. User-installed software should not go under `/usr`; they need to go under `/usr/local`.

stephanosio replied:
added `$HOME/.local` paths back in.

carlescufi commented on doc/getting_started/index.rst:
```diff
@@ -471,71 +471,142 @@ Install a Toolchain
 
-            chmod +x zephyr-sdk-0.13.2-linux-x86_64-setup.run
-            ./zephyr-sdk-0.13.2-linux-x86_64-setup.run -- -d ~/zephyr-sdk-0.13.2
+            tar xvf zephyr-sdk-0.14.0_linux-x86_64.tar.gz
```
Shouldn't users go to the folder first, before executing this command?

carlescufi replied:
Applies to macOS too

stephanosio replied:
That was written with the premise that the user is installing the Zephyr SDK under `~`.

Note that the previous step "Download and verify the latest Zephyr SDK bundle" already specifies `cd ~`.

stephanosio replied:
mainly because using `~/.local` for installing programs is arguably controversial and not really a "standard practice."

e.g. https://refspecs.linuxfoundation.org/FHS_3.0/fhs-3.0.html

carlescufi commented on doc/getting_started/index.rst:
```diff
@@ -471,71 +471,142 @@ Install a Toolchain
+
+         .. code-block:: bash
+
+            tar xvf zephyr-sdk-0.14.0_macos-x86_64.tar.gz
```
Same observation about `cd` before this

carlescufi commented on doc/getting_started/index.rst:
```diff
@@ -471,71 +471,142 @@ Install a Toolchain
+         .. code-block:: bash
+
+            cd zephyr-sdk-0.14.0
+            ./setup.sh
```
this works on macOS?

stephanosio replied:
Yes.

https://github.com/zephyrproject-rtos/sdk-ng/blob/6e7d5d6d7ff0cc774d77c50a3c32e3ad2be73ba2/scripts/template_setup_posix#L1

AFAIK, bash is included in all macOS distributions, although they might not be up-to-date.

carlescufi commented on doc/getting_started/index.rst:
```diff
@@ -471,71 +471,142 @@ Install a Toolchain
+
+         .. code-block:: bash
+
+            cd %HOMEPATH%
```
Please ask users to open a native terminal window like `cmd.exe` or `Console`. There should be a reference to this elsewhere in the Windows doc

stephanosio replied:
will do

stephanosio replied:
done

carlescufi commented on doc/getting_started/index.rst:
```diff
@@ -471,71 +471,142 @@ Install a Toolchain
+            It is recommended to extract the Zephyr SDK bundle at one of the following locations:
+
+            * ``%HOMEPATH%``
+            * ``%PROGRAMFILES%``
```
This will have spaces in the path, is this OK?

stephanosio replied:
Yes, the Zephyr SDK setup script handles spaces in its path without any issues.

As for the Zephyr build system, as far as the toolchain install path is concerned, it should be able to handle the spaces in its path: https://github.com/zephyrproject-rtos/zephyr/pull/42742#issuecomment-1039122128

carlescufi commented on doc/getting_started/installation_linux.rst:
```diff
@@ -210,87 +210,89 @@ environment is recommended to avoid interfering with your system Python.
-installed above are only needed for installing the SDK.
+Use of the Zephyr SDK is highly recommended and may even be required under
+certain conditions (e.g. running tests in QEMU for some architectures). Some
+of the dependencies installed above are only needed for installing the SDK.
```
which ones? `wget` and `unzip`? Why the need to mention this?

stephanosio replied:
I kept that sentence because it was already there; but, rethinking about it, it would be better to just remove it to avoid confusions.

stephanosio replied:
done

carlescufi reviewed (changes requested):
Great job, some changes requested. Also, can you please update the [Set up a toolchain](https://docs.zephyrproject.org/latest/guides/beyond-GSG.html#set-up-a-toolchain) section? This paragraph is now wrong:

>On Linux systems, you can install the [Zephyr SDK](https://docs.zephyrproject.org/latest/getting_started/installation_linux.html#zephyr-sdk) to get toolchains for all supported architectures. Otherwise, you can install other toolchains in the usual way for your operating system: with installer programs or system package managers, by downloading and extracting a zip archive, etc.

stephanosio commented on doc/getting_started/index.rst:
```diff
@@ -471,71 +471,142 @@ Install a Toolchain
 
+The Zephyr Software Development Kit (SDK) contains toolchains for each of
+Zephyr's supported architectures. It also includes additional host tools, such
+as custom QEMU and OpenOCD.
```
not sure if "executables" signifies anything here.

QEMU and OpenOCD are "additional host tools" and not necessarily just executables.

carlescufi replied:
well, you need to add something here, "as custom QEMU and OpenOCD" doesn't work in English? "as custom versions/buids of QEMU and OpenOCD?

stephanosio replied:
fixed

FrancescoSer commented on doc/guides/beyond-GSG.rst:
```diff
@@ -64,22 +64,20 @@ a cross-compiler and related tools which are different than the compiler
+You can install the :ref:`Zephyr SDK <zephyr_sdk>` to get toolchains for all
+supported architectures.  Otherwise, you can install other toolchains in the
+usual way for your operating system: with installer programs or system package
+managers, by downloading and extracting a zip archive, etc.
```
```suggestion
managers, by downloading and extracting a zip archive, and so on.
```

stephanosio replied:
done

FrancescoSer commented on doc/getting_started/installation_linux.rst:
```diff
@@ -210,87 +210,89 @@ environment is recommended to avoid interfering with your system Python.
 
-   this allow Zephyr to pick the right toolchain, while allowing multiple Zephyr
-   SDKs to be grouped together at a custom location.
+   This allows the Zephyr build system to choose the correct version of SDK,
```
```suggestion
   This allows the Zephyr build system to choose the correct version of the SDK,
```

stephanosio replied:
done

FrancescoSer reviewed (changes requested):
Hi! Proposing a few small changes to the text.

stephanosio commented on doc/getting_started/index.rst:
```diff
@@ -471,71 +471,142 @@ Install a Toolchain
+            You only need to run the setup script once after extracting the Zephyr SDK bundle.
 
-            You cannot move the SDK directory after you have installed it.
+            It may be necessary to re-run the setup script if you relocate the Zephyr SDK bundle
```
done

stephanosio commented on doc/getting_started/index.rst:
```diff
@@ -471,71 +471,142 @@ Install a Toolchain
+         .. note::
+            You only need to run the setup script once after extracting the Zephyr SDK bundle.
+
+            It may be necessary to re-run the setup script if you relocate the Zephyr SDK bundle
```
done

stephanosio commented on doc/getting_started/index.rst:
```diff
@@ -471,71 +471,142 @@ Install a Toolchain
 
-      Do not forget to set the required :ref:`environment variables <env_vars>`
-      (:envvar:`ZEPHYR_TOOLCHAIN_VARIANT` and toolchain specific ones).
+            It may be necessary to re-run the setup script if you relocate the Zephyr SDK bundle
```
done

stephanosio commented on doc/getting_started/installation_linux.rst:
```diff
@@ -210,87 +210,89 @@ environment is recommended to avoid interfering with your system Python.
-Use of the Zephyr SDK is optional, but recommended. Some of the dependencies
-installed above are only needed for installing the SDK.
+Use of the Zephyr SDK is highly recommended and may even be required under
+certain conditions (e.g. running tests in QEMU for some architectures). Some
```
done

stephanosio commented on doc/getting_started/installation_linux.rst:
```diff
@@ -210,87 +210,89 @@ environment is recommended to avoid interfering with your system Python.
-#. Download the `latest SDK
-   <https://github.com/zephyrproject-rtos/sdk-ng/releases>`_ as a
-   self-extracting installation binary:
+   If your host architecture is 64-bit ARM (e.g. Raspberry Pi), replace
```
done

stephanosio commented on doc/getting_started/installation_linux.rst:
```diff
@@ -210,87 +210,89 @@ environment is recommended to avoid interfering with your system Python.
+   If this fails, make sure Zephyr's dependencies were installed as described
+   in `Install Requirements and Dependencies`_.
 
 If you ever want to uninstall the SDK, just remove the directory where you
```
done

stephanosio commented on doc/getting_started/installation_linux.rst:
```diff
@@ -210,87 +210,89 @@ environment is recommended to avoid interfering with your system Python.
 
 If you ever want to uninstall the SDK, just remove the directory where you
-installed it.
+installed it. If you relocate the SDK directory, you need to re-run the setup
```
done

stephanosio commented on doc/getting_started/installation_linux.rst:
```diff
@@ -210,87 +210,89 @@ environment is recommended to avoid interfering with your system Python.
-   installation or set :envvar:`ZEPHYR_SDK_INSTALL_DIR` to point to the Zephyr
-   SDK installation folder.
+   If you install the Zephyr SDK outside any of these locations, then it is
+   required to register the Zephyr SDK in the CMake package registry by running
```
done

stephanosio commented on doc/getting_started/installation_linux.rst:
```diff
@@ -210,87 +210,89 @@ environment is recommended to avoid interfering with your system Python.
-   for example: ``ZEPHYR_SDK_INSTALL_DIR=/company/tools``
+   :envvar:`ZEPHYR_SDK_INSTALL_DIR` can also be used for pointing to a
+   directory containing multiple Zephyr SDKs, allowing for automatic toolchain
+   selection, for example: ``ZEPHYR_SDK_INSTALL_DIR=/company/tools``
```
done

gmarull commented on doc/getting_started/index.rst:
```diff
@@ -471,71 +471,154 @@ Install a Toolchain
+            wget https://github.com/zephyrproject-rtos/sdk-ng/releases/download/v0.14.0/zephyr-sdk-0.14.0_macos-x86_64.tar.gz
+            wget -O - https://github.com/zephyrproject-rtos/sdk-ng/releases/download/v0.14.0/sha256.sum | shasum --check --ignore-missing
+
+         If your host architecture is 64-bit ARM (Apple Silicon, also known as M1), replace
```
I'd personally remove M1 references, I'm pretty sure we'll have M2 or M1plus or whatever at some point.

gmarull replied:
I think just mentioning `Apple Sillicon` means we don't have to keep updating the document when, e.g., M2 is released.

carlescufi replied:
I agree with @gmarull, let's stick to "Apple Silicon". Please send a follow-up PR @stephanosio.

stephanosio replied:
Looking at https://www.apple.com/mac-mini/, they almost exclusively refer to their Arm-based Mac Minis as "M1" and there is not a single mention of the "Apple Silicon" on that page (from what I can see, they stopped using the "Apple Silicon" term, at least in the marketing materials); in that sense, I think it is helpful to keep that "aka. M1" for now.

When they release a new M series SoC, I will make sure to send a follow-up PR either generalising that phrase or just completely removing it.

gmarull reviewed (approved):
lgtm, one minor suggestion
